Source/WTF/wtf/TZoneMalloc.h

```diff
@@ -75,25 +121,69 @@
+// special class (e.g. those used with CompactPtr) allocators with FastMalloc fallback if TZoneHeap is enabled.
+#define WTF_MAKE_COMPACT_TZONE_ALLOCATED(name) MAKE_BTZONE_MALLOCED(name, Compact, FastFallback, WTF_NOEXPORT)
+#define WTF_MAKE_COMPACT_TZONE_ALLOCATED_EXPORT(name, exportMacro) MAKE_BTZONE_MALLOCED(name, Compact, FastFallback, exportMacro)
+
+// IsoHeap fallback allocators
+
+// class allocators with IsoHeap fallback if TZoneHeap is enabled.
+#define WTF_MAKE_TZONE_OR_ISO_ALLOCATED(name) MAKE_BTZONE_MALLOCED(name, NonCompact, IsoFallback, WTF_NOEXPORT)
+#define WTF_MAKE_TZONE_OR_ISO_ALLOCATED_EXPORT(name, exportMacro) MAKE_BTZONE_MALLOCED(name, NonCompact, IsoFallback, exportMacro)
+
+// template allocators with IsoHeap fallback if TZoneHeap is enabled.
+#define WTF_MAKE_TZONE_OR_ISO_ALLOCATED_TEMPLATE(name) MAKE_BTZONE_MALLOCED_TEMPLATE(name, NonCompact, IsoFallback, WTF_NOEXPORT)
+
+// special class (e.g. those used with CompactPtr) allocators with IsoHeap fallback if TZoneHeap is enabled.
 #define WTF_MAKE_COMPACT_TZONE_OR_ISO_ALLOCATED(name) \
     WTF_ALLOW_COMPACT_POINTERS; \
-    MAKE_BTZONE_MALLOCED(name, CompactTZoneHeap, WTF_NOEXPORT)
+    MAKE_BTZONE_MALLOCED(name, Compact, IsoFallback, WTF_NOEXPORT)
 #define WTF_MAKE_COMPACT_TZONE_OR_ISO_ALLOCATED_EXPORT(name, exportMacro) \
     WTF_ALLOW_COMPACT_POINTERS; \
-    MAKE_BTZONE_MALLOCED(name, CompactTZoneHeap, exportMacro)
-#define WTF_MAKE_TZONE_OR_ISO_NONALLOCATABLE(name) WTF_FORBID_HEAP_ALLOCATION
+    MAKE_BTZONE_MALLOCED(name, Compact, IsoFallback, exportMacro)
+
+// Template implementations for instantiating allocator template static / methods
+
+// template implementation to go with WTF_MAKE_TZONE_ALLOCATED_TEMPLATE and
+// WTF_MAKE_TZONE_ALLOCATED_TEMPLATE_EXPORT if TZoneHeap is enabled. This
+// should be added immediately after the template definition.
+#define WTF_MAKE_TZONE_ALLOCATED_TEMPLATE_IMPL(_templateParameters, _type) MAKE_BTZONE_MALLOCED_TEMPLATE_IMPL(_templateParameters, _type, FastFallback)
+
+// template implementation to go with WTF_MAKE_TZONE_OR_ISO_ALLOCATED_TEMPLATE
+// if TZoneHeap is enabled. This should be added immediately after the template definition.
+#define WTF_MAKE_TZONE_OR_ISO_ALLOCATED_TEMPLATE_IMPL(_templateParameters, _type) MAKE_BTZONE_MALLOCED_TEMPLATE_IMPL(_templateParameters, _type, IsoFallback)
+
+// template implementation for to go with WTF_MAKE_TZONE_ALLOCATED_TEMPLATE and
+// WTF_MAKE_TZONE_ALLOCATED_TEMPLATE_EXPORT if TZoneHeap is ensabled. This
+// should be added immediately after the template definition. This version is
+// needed in order to support templates with multiple parameters (which
+// WTF_MAKE_TZONE_ALLOCATED_TEMPLATE_IMPL cannot support).
+//
+// Requires the client to define these 3 macros:
+//     TZONE_TEMPLATE_PARAMS, TZONE_TYPE, TZONE_TYPE_NAME
+#define WTF_MAKE_TZONE_ALLOCATED_TEMPLATE_IMPL_WITH_MULTIPLE_OR_SPECIALIZED_PARAMETERS() \
+    MAKE_BTZONE_MALLOCED_TEMPLATE_IMPL_WITH_MULTIPLE_PARAMETERS()
```

Comment:
Where is `MAKE_BTZONE_MALLOCED_TEMPLATE_IMPL_WITH_MULTIPLE_PARAMETERS` defined?

Reply:
TZoneHeap.h.